File: erlang/etudes/ch3/geom.erl
%% @author Howard Fear 
%% @doc Functions calculating various geometric properties

-module(geom).
-export([area/1]).

%% @doc Calculates area of various shapes.

-spec area({atom(), number(), number()}) -> number().

area({Shape, A, B}) -> area(Shape, A, B).
area(rectangle, Length, Width) when Length > 0, Width > 0 -> Length * Width;
area(triangle, Base, Height) when Base > 0, Height > 0 -> (Base * Height) / 2;
area(ellipse, Major, Minor) when Major > 0, Minor > 0 -> math:pi() * Major * Minor;
area(_, _, _) -> 0.

File: elixer/etudes/ch03/geom.ex
defmodule Geom do
  def area({shape, a, b}) do
    area(shape, a, b)
  end
  defp area(:rectangle, height, width) when height >= 0 and width >= 0 do
    height * width
  end
  defp area(:triangle, base, height) when base >= 0 and height >= 0 do
    (base * height) / 2.0
  end
  defp area(:ellipse, major, minor) when major >= 0 and major >= 0 do
    :math.pi() * major * minor
  end
  defp area(_, _, _) do
    0
  end
end
